File: solutions/python/collatz-conjecture/1/collatz_conjecture.py
def steps(number):
    try:
        if number <= 0:
            raise  
        n = int(number)
    except:
        raise ValueError("Only positive integers are allowed")
    def fly_time(n):
        if n&1:
            if n == 1:
                return 0
            return 1 + fly_time(3*n+1)
        return 1 + fly_time(n//2)
    return fly_time(n)

File: solutions/javascript/collatz-conjecture/2/collatz-conjecture.js
export const steps = (n) => {
  // if (n > 1) return steps(n&1?3*n+1:n>>1)+1; 
  // if (n == 1) return 0;
  // throw Error('Only positive numbers are allowed');

  // non-recursive version for comparison purpose
  // looks like node correctly apply TCO
  // as recursive and non recursive version have
  // basically the same performance.
  for (var count = 0 ; n > 1 ; count++) {
    n = n&1?3*n+1:n>>1;
  }
  if (n == 1) return count;
  throw Error('Only positive numbers are allowed');
};
